ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -0,0 +1,339 @@
+    // externals will set the correct values for these two variables
+    state = State.FREE;
+    endOffset = new AtomicLong(0);
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * Attempts to write the {@code buffer} in its entirety in this segment. To guarantee that the write is persisted,
+   * {@link #flush()} has to be called.
+   * <p/>
+   * The write is not started if it cannot be completed.
+   * @param buffer The buffer from which data needs to be written from
+   * @return the number of bytes written.
+   * @throws IllegalArgumentException if there is not enough space for {@code buffer}
+   * @throws IllegalStateException if the segment is not {@link State#ACTIVE}.
+   * @throws IOException if data could not be written to the file because of I/O errors
+   */
+  @Override
+  public int appendFrom(ByteBuffer buffer)
+      throws IOException {
+    int bytesWritten = 0;
+    ensureActive();
+    if (endOffset.get() + buffer.remaining() > capacityInBytes) {
+      metrics.overflowWriteError.inc();
+      throw new IllegalArgumentException(
+          "Buffer cannot be written to segment [" + file.getAbsolutePath() + "] because " + "it exceeds the capacity ["
+              + capacityInBytes + "]");
+    } else {
+      while (buffer.hasRemaining()) {
+        bytesWritten += fileChannel.write(buffer, endOffset.get());
+      }
+      endOffset.addAndGet(bytesWritten);
+    }
+    return bytesWritten;
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * Attempts to write the {@code channel} in its entirety in this segment. To guarantee that the write is persisted,
+   * {@link #flush()} has to be called.
+   * <p/>
+   * The write is not started if it cannot be completed.
+   * @param channel The channel from which data needs to be written from
+   * @param size The amount of data in bytes to be written from the channel
+   * @throws IllegalArgumentException if there is not enough space for data of size {@code size}.
+   * @throws IllegalStateException if the segment is not {@link State#ACTIVE}.
+   * @throws IOException if data could not be written to the file because of I/O errors
+   */
+  @Override
+  public void appendFrom(ReadableByteChannel channel, long size)
+      throws IOException {
+    ensureActive();
+    if (endOffset.get() + size > capacityInBytes) {
+      metrics.overflowWriteError.inc();
+      throw new IllegalArgumentException(
+          "Channel cannot be written to segment [" + file.getAbsolutePath() + "] because" + " it exceeds the capacity ["
+              + capacityInBytes + "]");
+    } else {
+      long bytesWritten = 0;
+      while (bytesWritten < size) {
+        bytesWritten += fileChannel.transferFrom(channel, endOffset.get() + bytesWritten, size - bytesWritten);
+      }
+      endOffset.addAndGet(bytesWritten);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * The read is not started if it cannot be completed.
+   * @param buffer The buffer into which the read needs to write to
```

Comment:
this is derived from the `Read` interface where it makes sense. I have changed this to "The buffer into which the data needs to be written"